_posts/2018-07-25-dota-macos-performance.md

```diff
@@ -0,0 +1,83 @@
+---
+layout: post
+title: Portability benchmark of Dota2 on MacOS
+---
+
+### The Race
+
+gfx-rs is a Rust project aiming to make graphics programming more accessible and portable, focusing on Vulkan. It's a single Rust API with multiple backends: D3D12/11, Metal, Vulkan, and even OpenGL. We are also building a Vulkan Portability [implementation](https://github.com/gfx-rs/portability) based on it, which allows non-Rust applications using Vulkan to run everywhere.
+
+As our portability implementation was maturing in terms of [CTS](https://github.com/KhronosGroup/VK-GL-CTS) coverage, we started to ask questions about performance. Fortunately, just around this time frame [Valve announced](https://twitter.com/plagman2/status/1002324195135520768?lang=en) official Metal support for Dota2, via [MoltenVK](https://github.com/KhronosGroup/MoltenVK). We couldn't miss the opportunity to test _gfx-portability_ on Dota2, and replacing MoltenVK appeared to be straightforward. The initial results were... slightly discouraging. For a while we weren't able to get the game to start, since our Metal backend was missing some required features, like compressed texture support.
+
+Then we entered a long period of mixed stability and performance fixes. We had to rethink the state caching architecture, resource lifetimes, lock contention, and much more. Finally, we figured out all the issues we had with Apple, Metal, and Dota itself and were ready for the race.
+
+![simple test with gfx-portability](/img/dota-simple.jpg)
+
+#### Modes
+
+Our portability library has been tested in two different modes:
```

Comment:
These modes are only relevant to the Metal backend and not across all of portability